projects/ledger-app/0001-token-admin-methods.md

```diff
@@ -0,0 +1,219 @@
+Since the Ledger cannot check if a token is an NFT or not, we should disregard any changes and treat both as custom tokens (which is valid if we follow the protocol).
+
+## Transaction parsing
+
+To allow admin operations we will introduce new concepts into the app that were being ignored or not treated.
+
+---
+
+The `token_data` of an output is a byte with the information:
+- Weather the token is a custom token or HTR
+- Which custom token it is (index on the token array) 
+- If the output is an authority output (which authority will be encoded on the `value`)
+Currently on the app we ignore the last piece of information since we do not allow authority outputs.
+The way we do it is by throwing an error of the `token_data & TOKEN_DATA_AUTHORITY_MASK` is not 0.
+To allow authority outputs we just need to change this and create a special screen for authority outputs when the user is confirming them.
+Usually the output confirmation screen involves 4 steps:
+- Address step (with pagination)
+- Amount step (Showing as `ABC 0.01` or `HTR 0.01`, with pagination for big values)
+- Approve step
+- Reject step
+
+When confirming an authority output we should show:
+- Address step (with pagination)
+- Type step (`mint authority` or `melt authority`)
+	- Other authorities will throw.
+	- We will not allow joined authority, for instance `mint+melt authority` (i.e. `value = 3`)
+- Approve step
+- Reject step
+
+Obs: With authority outputs we already have enough to enable mint, melt, delegate and destroy operations.
+
+---
+
+Currently we only allow P2PKH scripts (and addresses) but NFTs have a data output.
+Data outputs are allowed on the protocol but we will only allow on NFT creation scripts because any other is just an HTR burn.
+Data outputs are allowed to use custom tokens on the protocol but we will only allow HTR (token index 0) since NFT creation requires HTR.
+
+The user will be required to confirm the data on the script, which means that only ascii characters will be allowed since Ledger does not support UTF-8.
+
+---
+
+When creating a custom token or NFT the transaction information also includes the name and symbol of the token, these should only exist when a transaction has the version byte 2.
+
+---
+
+> [!NOTE] Version field has been split into `signalBits` and `version`
+> The version field of the transaction was 2 bytes (first byte was reserved for later use) and since Ledger app version 1.1.0 the version field has become a 1 byte field and the first byte became the `signalBits`
+
+Version byte needs to be refactored to a 1 byte field and support for `signalBits` will not be added in this design so we will only allow `0x00` as `signalBits` for now.
+
+## Admin operations
+
+### Mint, Melt, Delegate and Destroy
+
+Sending a mint, melt, delegate or destroy transactions should use the same protocol as sending normal transactions.

```

Comment:
How are we going to present which is the token id that is being minted/melted/delegated/destroyed? I feel that we should present the Token Name, Token Symbol and Token UID.

I also think that the wallet should send the TokenCreationTransaction to the Ledger app, so it doesn't have to trust metadata. Then, the Ledger app will calculate the hash, confirm it is the same as the token_id in the new transaction, and, finally, get the token name and token symbol from the TokenCreationTransaction.

Reply:
> How are we going to present which is the token id that is being minted/melted/delegated/destroyed? I feel that we should present the Token Name, Token Symbol and Token UID.

Since the Ledger does not have the transaction from the inputs we are somewhat limited here.
Having all spent transactions sent to Ledger is not feasible either since it may not have enough memory for this (nanoS has about 4.5Kb of RAM).
So a mint operation that does not create a new mint authority is indistinguishable (from the outputs only) from a transaction that is sending some tokens, we may add some information to make the UX nicer, but the ledger may not be able to check that mint/melt/destroy operations are being done (delegate operations always have an authority output).
An example of this is that if we do not send authority outputs we can already mint and melt tokens and destroy token authorities with the Ledger device.

> I also think that the wallet should send the TokenCreationTransaction to the Ledger app, so it doesn't have to trust metadata. Then, the Ledger app will calculate the hash, confirm it is the same as the token_id in the new transaction, and, finally, get the token name and token symbol from the TokenCreationTransaction.

Currently the user has to manually confirm any tokens that the he wants to send in transactions with the Ledger, this process is called token signing and must be done before we attempt to sign a transaction with the token (the app will block a transaction if a token in the token array was not signed by the user).
We could add the "TokenCreationTransaction hash matching the UID" check to this token signing to make it even safer, but I think we should avoid blindly trusting a token to be sent using the Ledger just because it matches the TokenCreationTransaction.
Also, I think this would merit a design of its own.